src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs

```diff
@@ -58,28 +55,35 @@ public static partial class JsonSerializer
 
             if (state.Current.CollectionEnumerator.MoveNext())
             {
+                // A dictionary should not have a null KeyValuePair.
+                Debug.Assert(state.Current.CollectionEnumerator.Current != null);
+
+                bool obtainedValues = false;
+                string key = default;
+                object value = default;
+
                 // Check for polymorphism.
                 if (elementClassInfo.ClassType == ClassType.Unknown)
                 {
-                    object currentValue = ((IDictionaryEnumerator)state.Current.CollectionEnumerator).Entry.Value;
-                    GetRuntimeClassInfo(currentValue, ref elementClassInfo, options);
+                    jsonPropertyInfo.GetDictionaryKeyAndValue(ref state.Current, out key, out value);
+                    GetRuntimeClassInfo(value, ref elementClassInfo, options);
+                    obtainedValues = true;
                 }
 
                 if (elementClassInfo.ClassType == ClassType.Value)
                 {
                     elementClassInfo.PolicyProperty.WriteDictionary(ref state, writer);
```

Comment:
Since we might have the `value` here from the polymorphic check above, should we have logic here to see if we do, and just write the `key` and `value` here, skipping all the logic that will happen in (and on the path to) `WriteDictionary<TProperty>()` below?

Reply:
Yes passing the key\value down is possible here but requires a new helper method that would essentially copy most of the code\flow in WriteDictionary in order to bounce from non-generic to generic. Plus add associated tests.

Since the original code also obtained the key and value before calling WriteDictionary, which didn't change here, I'll include this optimization in the other PR mentioned earlier for boxing. Thanks